### PR TITLE
Fix: range control direct entry in input field

### LINKED ISSUE #24460: Spacer Block: Difficult to enter values directly into input field
**Is your feature request related to a problem? Please describe.**
Yes. If I try to empty the input field to directly put my height numbers into it (via num keys), it jumps back to 20px.
This isn't very convenient. **Test:**

1. Create a spacer block.
2. Mark the number in the input field.
3. Try to write like 500 via num keys. The field jumps to 200.
(Or try to change the height from 20 to 40 only with keyboard input, which is a common task on daily basis.)

**Describe the solution you'd like**
As long I have focus on the field it shouldn't update itself to the min height.
Currently it's updating on input change.

If I don't set any number or a number below min height and focus is lost, it should update itself to the min height.

**Describe alternatives you've considered**
There is no other way, just don't block direct keyboard input for number input fields.

### PR DESCRIPTION
**Pull request by stokesman**
The impetus is to fix #24460 which is a [regression](21549). The solution in this PR is simple and includes a test to prevent regressions but did require minor changes to quite a few tests.  Also, though tangential to the issue, there is the removal of some extraneous code.

The core change is to InputControl, revising logic for value state updates and propagation. It no longer updates the internal value state from props while the input has focus. The change ensures that entry will be unimpeded by the validation logic of consuming components (such as RangeControl).

A new test is added for RangeControl specifying the allowance of invalid values in the input element until the loss of focus. It was confirmed to fail before the rest of the changes were added and so guards against future regressions. 

Ancillary changes are the removal of RangeControl's wrapping of NumberControl and test updates to work with the new InputControl implementation detail. 

## How has this been tested?
- Storybook
- WordPress 5.5.1
- Unit tests

## Types of changes
- Fixes: #24460, which despite the title affects RangeControl itself and all consumers of it.
- Revised InputControl value state synchronization logic
- Removal of extraneous wrapper around NumberControl in RangeControl
- Added one test for RangeControl and changed one other.
- Updated all tests involving InputControl, improving some by converting them to use react-testing-library.
- Updated AnglePickerControl to avoid issue with new InputControl implementation detail

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [n/a] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [n/a] I've included developer documentation if appropriate. 
- [n/a?] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/input-control/input-field.js [hunk 1]
```diff
@@ -63,35 +66,27 @@ function InputField(
...
-		 * This effectively enables a "controlled" state.
-		 * https://reactjs.org/docs/forms.html#controlled-components
-		 */
-		if ( valueProp !== valueRef.current ) {
-			update( valueProp );
-			valueRef.current = valueProp;
-
-			// Quick return to avoid firing the onChange callback
+	/*
+	 * Syncs value state using the focus state to determine the direction.
```
Comment (ItsJonQ) [inline packages/components/src/input-control/input-field.js:72]: @stokesman Interesting strategy! This simplifies the syncing a lot 👏

File: packages/components/src/range-control/index.js [hunk 2]
```diff
@@ -256,16 +274,19 @@ function RangeControl(
 					</AfterIconWrapper>
 				) }
 				{ withInputField && (
-					<InputField
+					<InputNumber
+						aria-label={ label }
+						className="components-range-control__number"
 						disabled={ disabled }
+						inputMode="decimal"
```
Comment (ItsJonQ) [inline packages/components/src/range-control/index.js:281]: @stokesman Nice enhancement! Thank you for that 🙏

File: packages/components/src/range-control/index.js [hunk 3]
```diff
@@ -256,16 +274,19 @@ function RangeControl(
...
 						isShiftStepEnabled={ isShiftStepEnabled }
-						label={ label }
 						max={ max }
 						min={ min }
+						onBlur={ handleOnInputNumberBlur }
 						onChange={ handleOnChange }
-						onReset={ handleOnReset }
 						shiftStep={ shiftStep }
 						step={ step }
+						type={ 'number' }
```
Comment (ItsJonQ) [inline packages/components/src/range-control/index.js:289]: @stokesman The `type` should be `number` by default for `NumberControl`. I don't think we need this :)

**Review by ItsJonQ — APPROVED** [review]
@stokesman Wowow! Thank you so much for improving this. I had tried to improve the `value` sync (between uncontrolled and controlled) in the latest iterations of these controls (the base `InputControl`).

I think you cracked it by leveraging the `blur` event as the "trigger" mechanism to consolidate - very clever 👏 !!!

I left a comment in regards to the `type` for `NumberControl` within `RangeControl`. Not a huge thing at all.

Tested all of the affected controls locally.
Tested mouse + keyboard interactions as well as controlled and uncontrolled states.
It's working as expected!

Approved from me!

P.S. A part of me is still unsure about using the `blur` approach, but that's 100% due to my personal unfamiliarity with the technique - rather than something being wrong <3

File: packages/components/src/range-control/index.js (hunk 3 above)
Comment (stokesman) [reply packages/components/src/range-control/index.js]: Quite right, it's gone!

**Comment by stokesman** [comment]
I've pushed some updates because I found an issue with the value reset on blur happening when it shouldn't. Along with that, I realized I had some extraneous logic in there and what was probably overeager avoidance of `onChange` for out-of-range values. Also I've updated the readme.

@ItsJonQ, I'm glad you were added to review.

> P.S. A part of me is still unsure about using the blur approach

I very much relate to that! I've gotten to be fairly confident it's good for this use case. I was also unsure about putting it in `InputControl` but besides how it simplied the sync logic, liked that it allowed removing the added layer of state management for the number input in `RangeControl`. 

Speaking of that, I've come to suspect the value state management for `RangeControl` is vestigal from before `NumberControl` was integrated. It seems that `RangeControl` would be perfectly fine to always be "controlled" now. I've given it a quick test as such and had success. Any thoughts on that? Not that we'd need to add it here.

**Comment by ItsJonQ** [comment]
@stokesman I've tested it on my side, and it works as expected :).

The changes look good to me 👍 . If something unexpected just arise from this update, we can refine the implementation then! (or revert if we need and revisit how state syncing should be handled, haha)

Thanks again!

**Comment by jasmussen** [comment]
💯

**Pull request merged**